File: GO/a2zSheet/step17_tries/prefix_trie.cpp
#include <bits/stdc++.h> 

struct Node{
    Node* links[26];
    int ends = 0;
    int pref = 0;
};

class Trie{
    private: Node* root;
    public:

    Trie(){
        root = new Node();
    }

    void insert(string &word){
        Node* temp = root;
        for(auto &c:word){
            if(!temp->links[c-'a']){
                temp->links[c-'a'] = new Node();
            }
            temp = temp->links[c-'a'];
            temp->pref++;
        }

        temp->ends++;

    }

    int countWordsEqualTo(string &word){
        Node* temp = root;
        for(auto &c:word){
            if(temp->links[c-'a']){
                temp = temp->links[c-'a'];
            } else return 0;
        }

        return temp->ends;
    }

    int countWordsStartingWith(string &word){
        Node* temp = root;
        for(auto &c:word){
            if(temp->links[c-'a']){
                temp = temp->links[c-'a'];
            }
            else return 0;
            
        }

        return temp->pref;
    }

    void erase(string &word){
        Node* temp = root;
        for(auto &c:word){
            if(temp->links[c-'a']){
                temp = temp->links[c-'a'];
            }
            temp->pref--;
            
        }
        temp->ends--;
    }
};

File: a2zSheet/step17_tries/prefix_trie.go
type Node struct {
	links [26]*Node
	pref  int
	ends  int
}

type Trie struct {
	root *Node
}

func Constructor() Trie {
	return Trie{root: &Node{}}
}

func (t *Trie) insert(word string) {
	temp := t.root
	for _, c := range word {
		if temp.links[c-'a'] == nil {
			temp.links[c-'a'] = &Node{}
		}
		temp = temp.links[c-'a']
		temp.pref++
	}
	temp.ends++
}

func (t *Trie) countWordsEqualTo(word string) int {
	temp := t.root
	for _, c := range word {
		if temp.links[c-'a'] != nil {
			temp = temp.links[c-'a']
		} else {
			return 0
		}
	}
	return temp.ends
}

func (t *Trie) countWordsStartingWith(word string) int {
	temp := t.root
	for _, c := range word {
		if temp.links[c-'a'] != nil {
			temp = temp.links[c-'a']
		} else {
			return 0
		}
	}
	return temp.pref
}

func (t *Trie) erase(word string) {
	temp := t.root
	for _, c := range word {
		if temp.links[c-'a'] != nil {
			temp = temp.links[c-'a']
			temp.pref--
		}
	}
	temp.ends--
}
